wordpress-mobile/release-toolkit
# Add some unit tests

## What is this about

As promised in https://github.com/wordpress-mobile/release-toolkit/pull/177#pullrequestreview-545829378:

* Adding some tests to our Fastlane actions, especially the new linter action
* Updated some of our gems in `Gemfile.lock` in the process

## How

Added a spec file for the linter spec + `yaml` files to provide definition of fixtures data + expected result. The tests mainly read the YAML file which tells it how to set up a test temporary directory with dummy `.lproj` and `.strings` files with the appropriate contents, then run the linter, then check that the linter result (returned hash) matches the expectation as defined by the YAML.

Note: I chose YAML over JSON mainly because of its ability to write strings directly without the need to quote them – which would otherwise have required a LOT of backslash-escapes in a JSON file to represent the same content.

_Note: Encountered a challenge with Fastlane::Actions subtle internal tweaks that makes it behave differently when running under tests. Especially the `Action.sh` refused to execute when it was run in the context of a unit test, while it worked as expected otherwise. Turns out, Fastlane has a test hardcoded in the implementation of `Actions.sh`, which changes its behavior under tests… but I had to dig quite a bit to find that out 😰 . I included a link to the relevant part in Fastlane's code as a code comment for reference for later._

## To test

* Clone this branch
* Run `rbenv exec bundle install` (I took the occasion to update some gems to latest version)
* Run `rbenv exec bundle exec rspec -fd` (`-fd` option is to format the test results with a more verbose output so you can see the name of each test that is run, and ensure our new tests are part of the set)
* Check that all the tests, and especially the ones about `Fastlane::Actions::IosLintLocalizationsAction`, are all green.

```
Fastlane::Actions::IosLintLocalizationsAction
  SwiftGen Installation Logic
    only installs SwiftGen the first time, when it is not yet installed
  Linter
    succeeds when there are no violations
    detects inconsistent placeholder count
    detects inconsistent placeholder types
    properly handles misleading characters and placeholders in RTL languages
    does not fail if a locale does not have any Localizable.strings
```

OR… [trust CI being all green](https://app.circleci.com/pipelines/github/wordpress-mobile/release-toolkit/466/workflows/98fe951a-ae5e-410f-9518-0dd1b3a9fd24/jobs/451) when it ran rspec 😛
